packages/SwingSet/src/kernel/vat-warehouse.js

```diff
@@ -38,8 +42,31 @@ function recordSyscalls(origHandler) {
     syscalls.push({ s: vso, r: vres });
     return vres;
   };
-  const getTranscriptSyscalls = () => syscalls;
-  return { syscallHandler, getTranscriptSyscalls };
+  const getTranscriptEntry = (vd, deliveryResult) => {
+    // TODO add metering computrons to results
+    /** @type {TranscriptDeliveryResults} */
+    const tdr = { status: deliveryResult[0] };
+    const transcriptEntry = { d: vd, sc: syscalls, r: tdr };
+    return transcriptEntry;
+  };
+  return { syscallHandler, getTranscriptEntry };
+}
+
+/**
+ * @param {TranscriptEntry} transcriptEntry
+ * @returns {VatDeliveryObject}
+ */
+function onlyRealDelivery(transcriptEntry) {
+  const dtype = transcriptEntry.d[0];
+  if (
+    dtype === 'save-snapshot' ||
+    dtype === 'load-snapshot' ||
+    dtype === 'initialize-worker' ||
+    dtype === 'shutdown-worker'
```

Comment:
Might it be safer under maintenance to look for and then allow the kinds of entries that are allowed under replay rather than looking for and then rejecting the ones that are prohibited?  It's one less special `if` that we need to remember to update if we add another special entry type.

Reply:
Hm. The list of ordinary deliveries lives in the `VatDeliveryObject` type in liveslots:

https://github.com/Agoric/agoric-sdk/blob/dde6eee8eb84b930062c962648161205520ff0dc/packages/swingset-liveslots/src/types.js#L34-L47

and currently contains 9 types: `message`, `notify`, `dropExports`, `retireExports`, `retireImports`, `changeVatOptions`, `startVat`, `stopVat`, `bringOutYourDead`. To do it right, we'd need liveslots to export a (non-type) list of normal ones. Except that different vats can be using different versions of liveslots, so if we e.g. add one in the future, then this vat-warehouse.js check would need to fetch the right list (`vatKeeper.getLiveslotsSettings`??).

It feels a bit more future-proof to have swingset reject the pseudo-delivery types that swingset has defined (since no version of liveslots will know about any of these), than to have it only accept the normal-delivery types that liveslots (which version?) defines.

But it *would* feel a bit better if swingset had a concrete list of these pseudo-delivery types, rather than vat-warehouse having a special `if` statement that is supposed to know about them. I'm not sure where that list ought to live, though, they're nominally defined in `types-internal.js`:

https://github.com/Agoric/agoric-sdk/blob/b91f3ad4c5ce51bc7c1fe6fa9a11afb1d7e335a2/packages/SwingSet/src/types-internal.js#L102-L112

but of course types aren't available for introspection by code. I suppose I could add a real export to `types-internal.js`, with a list of names? Or a `Set`, which might allow for faster lookup?

Anyways, I'm inclined to defer this to the future.

Reply:
Couldn't there simply be an `isDeliverable` predicate?

Reply:
If `isDeliverable` would live in `vat-warehouse.js`, then that's basically what `onlyRealDelivery` is, except that it takes a transcriptEntry and returns a delivery object:
```js
/**
 * @param {TranscriptEntry} transcriptEntry
 * @returns {VatDeliveryObject}
 */
function onlyRealDelivery(transcriptEntry) {
  const dtype = transcriptEntry.d[0];
  if (
    dtype === 'save-snapshot' ||
    dtype === 'load-snapshot' ||
    dtype === 'initialize-worker' ||
    dtype === 'shutdown-worker'
  ) {
    throw Fail`replay should not see ${dtype}`;
  }
  return transcriptEntry.d;
}
```

If it should live in liveslots, then we have the question of *which* (version of) liveslots it lives in. The kernel doesn't talk directly to liveslots, it talks to a vat which may or may not contain (some version of) liveslots. So there's no simple way for `packages/liveslots` to export a predicate that the kernel can query (no liveslots code is ever evaluated inside the kernel's Compartment).

Reply:
I would expect this to be a kernel predicate, as it's about purely kernel-side stuff that the kernel wants to avoid delivering to the vat.  My concern was simply that an allow-list is safer than a deny-list if we introduce another kernel-side special pseudo-deliverable thingie in the future.  If somebody does that and misses the need to update the `onlyRealDelivery` function here, then we could end up dropping something into a vat that could break it.  The list of things that _can_ be delivered into a vat is likely to be more stable, as there's no way to introduce a new actual deliverable type without coordination with the vat anyway.

To reiterate, this is purely a hygiene thing and doesn't touch on fundamental semantics.